### PR TITLE
DEV-411 staking reward - payout module extraction

### PR DESCRIPTION
**Pull request by TonySemikin**
#### Pre-Release

- [x] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [x] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/payment/models/payout/strategies/payout-token.strategy.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import { DexService } from '../../dex/services/dex.service';
+import { PayoutOrder, PayoutOrderContext } from '../entities/payout-order.entity';
+import { PayoutOrderRepository } from '../repositories/payout-order.repository';
+import { PayoutDeFiChainService } from '../services/payout-defichain.service';
+import { PayoutStrategy } from './payout.strategy';
+
+type TokenName = string;
+
+@Injectable()
+export class PayoutTokenStrategy extends PayoutStrategy {
```
Comment (davidleomay) [inline src/payment/models/payout/strategies/payout-token.strategy.ts:13]: Same question as last time 😉 I really like the strategy thinking ... but again there is a lot of duplicated code. E.g. the `sendToken` and `sendDFI` methods are practically the same. And it's not some trivial code but a very fundamental logic.
Is there a way to combine it? 😊

File: src/payment/models/payout/services/payout-defichain.service.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  async checkPayoutCompletion(context: PayoutOrderContext, payoutTxId: string): Promise<boolean> {
+    const transaction = await this.getClient(context).getTx(payoutTxId);
+
+    return transaction && transaction.blockhash && transaction.confirmations > 0;
+  }
+
+  async getUtxoForAddress(address: string): Promise<string> {
+    return this.whaleService.getClient().getBalance(address);
```
Comment (davidleomay) [inline src/payment/models/payout/services/payout-defichain.service.ts:35]: Convert to number?

File: src/payment/models/payout/services/payout-defichain.service.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  getWallet(context: PayoutOrderContext): string {
```
Comment (davidleomay) [inline src/payment/models/payout/services/payout-defichain.service.ts:38]: `getWalletAddress`?

File: src/payment/models/payout/strategies/payout.strategy.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    const uniqueAddresses = new Map<string, number>();
+
+    orders.forEach((o) => {
+      const existingAmount = uniqueAddresses.get(o.destinationAddress);
+      const increment = existingAmount ? Util.round(existingAmount + o.amount, 8) : o.amount;
+
+      uniqueAddresses.set(o.destinationAddress, increment);
+    });
+
+    return [...uniqueAddresses.entries()].map(([addressTo, amount]) => ({ addressTo, amount }));
```
Comment (davidleomay) [inline src/payment/models/payout/strategies/payout.strategy.ts:68]: Why this `[...uniqueAddresses.entries()]`? (same for other occasions in this file)

File: src/payment/models/buy-crypto/services/buy-crypto-out.service.ts [hunk 5]
```diff
@@ -77,181 +53,80 @@ export class BuyCryptoOutService {
...
 
-            console.info(`Paying out ${group.length} transaction(s). Transaction ID(s): ${group.map((t) => t.id)}`);
-
-            batch.outputAsset === 'DFI' ? await this.sendDFI(group) : await this.sendToken(group, batch.outputAsset);
-          } catch (e) {
-            console.error(
-              `Failed to payout group of ${group.length} transaction(s). Transaction ID(s): ${group.map((t) => t.id)}`,
-            );
+            console.error(`Failed to initiate buy-crypto payout. Transaction ID: ${transaction.id}`);
             // continue with next group in case payout failed
```
Comment (davidleomay) [inline src/payment/models/buy-crypto/services/buy-crypto-out.service.ts:79]: Comment needs to be updated 😉

**Review by davidleomay — COMMENTED** [review]
Generally, it looks very good 👏  I'm missing a bit of an overview, as it includes quite a few layers of abstraction. Do you have a drawing of all concerned services? Or would you mind setting up something?

File: src/payment/models/payout/services/payout.service.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+      if (existingOrder) {
+        throw new DuplicatedEntryException(
+          `Payout order for context ${context} and correlationId ${correlationId} already exists. Order ID: ${existingOrder.id}`,
+        );
+      }
+
+      const order = this.payoutOrderFactory.createOrder(request, Blockchain.DEFICHAIN);
+
+      await this.payoutOrderRepo.save(order);
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/payment/models/payout/services/payout.service.ts:53]: I wanted to make creation of payout order very safe and simple, it's only one DB write, so its acid. If it's more than one write that's already distributed transaction of some sort. imho in our case simplicity, testability and extendability beats performance penalty.

File: src/payment/models/payout/services/payout.service.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    const orders = await this.payoutOrderRepo.find({ status: PayoutOrderStatus.CREATED });
+    const confirmedOrders = [];
+
+    for (const order of orders) {
+      try {
+        const { asset, amount, context } = order;
+        const destinationAddress = this.defichainService.getWallet(context);
+        const request = { asset, amount, destinationAddress };
+
+        const transferTxId = await this.dexService.transferLiquidity(request);
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/payment/models/payout/services/payout.service.ts:150]: cool, then dex also have to be made configurable for node and payout need to have a separate strategy for direct payout w/o additional transfer

File: src/payment/models/payout/strategies/payout.strategy.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  protected aggregatePayout(orders: PayoutOrder[]): { addressTo: string; amount: number }[] {
+    // sum up duplicated addresses, fallback in case orders to same address and asset end up in one payment round
+    const uniqueAddresses = new Map<string, number>();
+
+    orders.forEach((o) => {
+      const existingAmount = uniqueAddresses.get(o.destinationAddress);
+      const increment = existingAmount ? Util.round(existingAmount + o.amount, 8) : o.amount;
+
+      uniqueAddresses.set(o.destinationAddress, increment);
+    });
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/payment/models/payout/strategies/payout.strategy.ts:66]: will try it out

File: src/payment/models/payout/strategies/payout-token.strategy.ts (hunk 1 above)
Comment (TonySemikin) [reply src/payment/models/payout/strategies/payout-token.strategy.ts:13]: hmm, was trying to do that already and put everything shared in superclass, but totally open to suggestions, lets discuss this

File: src/payment/models/payout/services/payout-defichain.service.ts (hunk 2 above)
Comment (TonySemikin) [reply src/payment/models/payout/services/payout-defichain.service.ts:35]: totally!

File: src/payment/models/payout/services/payout-defichain.service.ts (hunk 3 above)
Comment (TonySemikin) [reply src/payment/models/payout/services/payout-defichain.service.ts:38]: makes sense!

File: src/payment/models/payout/strategies/payout.strategy.ts (hunk 4 above)
Comment (TonySemikin) [reply src/payment/models/payout/strategies/payout.strategy.ts:68]: cause `.entries()` return Iterable and not Array, so I wouldn't be able to call `.map()` on it

File: src/payment/models/buy-crypto/services/buy-crypto-out.service.ts (hunk 5 above)
Comment (TonySemikin) [reply src/payment/models/buy-crypto/services/buy-crypto-out.service.ts:79]: yep

**Pull request closed without merging**